background/usercss-helper.js

```diff
@@ -1,42 +1,66 @@
+/* exports usercssHelper */
 'use strict';
 
-(() => {
+// eslint-disable-next-line no-unused-vars
+const usercssHelper = (() => {
+  const installCodeCache = {};
+  const clearInstallCode = url => delete installCodeCache[url];
+  const isResponseText = r => /^text\/(css|plain)(;.*?)?$/i.test(r.headers.get('content-type'));
+  // detecting FF68 by the added feature as navigator.ua may be spoofed via about:config or devtools
+  const fileLoader = !chrome.app && chrome.storage.managed &&
+    (tabId => browser.tabs.executeScript(tabId, {file: '/content/download-self.js'}).then(r => r[0]));
+
   API_METHODS.installUsercss = installUsercss;
   API_METHODS.editSaveUsercss = editSaveUsercss;
   API_METHODS.configUsercssVars = configUsercssVars;
 
   API_METHODS.buildUsercss = build;
-  API_METHODS.openUsercssInstallPage = install;
-
   API_METHODS.findUsercss = find;
 
-  const TEMP_CODE_PREFIX = 'tempUsercssCode';
-  const TEMP_CODE_CLEANUP_DELAY = 60e3;
-  let tempCodeLastWriteDate = 0;
-  if (FIREFOX) {
-    // the temp code is created on direct installation of usercss URLs in FF
-    // and can be left behind in case the install page didn't open in time before
-    // the extension was updated/reloaded/disabled or the browser was closed
-    setTimeout(function poll() {
-      if (Date.now() - tempCodeLastWriteDate < TEMP_CODE_CLEANUP_DELAY) {
-        setTimeout(poll, TEMP_CODE_CLEANUP_DELAY);
-        return;
+  API_METHODS.getUsercssInstallCode = url => {
+    // when the installer tab is reloaded after the cache is expired, this will throw intentionally
+    const {code, timer} = installCodeCache[url];
+    clearInstallCode(url);
+    clearTimeout(timer);
+    return code;
+  };
+
+  return {
+
+    testUrl(url) {
+      return url.includes('.user.') &&
+        /^(https?|file|ftps?):/.test(url) &&
+        /\.user\.(css|styl)$/.test(url.split(/[#?]/, 1)[0]);
+    },
+
+    /** @return {Promise<{ code:string, inTab:boolean } | false>} */
+    testContents(tabId, url) {
+      const isFile = url.startsWith('file:');
+      const inTab = isFile && Boolean(fileLoader);
+      return Promise.resolve(isFile || fetch(url, {method: 'HEAD'}).then(isResponseText))
+        .then(ok => ok && (inTab ? fileLoader(tabId) : download(url)))
+        .then(code => /==userstyle==/i.test(code) && {code, inTab});
+    },
+
+    openInstallerPage(tabId, url, {code, inTab} = {}) {
+      const newUrl = `${URLS.installUsercss}?updateUrl=${encodeURIComponent(url)}`;
+      if (inTab) {
+        getTab(tabId).then(tab =>
+          openURL({
+            url: `${newUrl}&tabId=${tabId}`,
+            active: tab.active,
+            index: tab.index + 1,
+            openerTabId: tabId,
+            currentWindow: null,
+          }));
+      } else {
+        const timer = setTimeout(clearInstallCode, 10e3, url);
+        installCodeCache[url] = {code, timer};
+        chrome.tabs.update(tabId, {url: newUrl});
```

Comment:
We can use `updateTab` here:
```js
updateTab(tabId, {url: newUrl}).catch(console.error);
```

Reply:
Why do you add a catch here? It should be printed to the console by default as a noncaught exception.